pkg/cortex/modules.go

```diff
@@ -614,3 +637,25 @@ var modules = map[ModuleName]module{
 		deps: []ModuleName{Querier, Ingester, Distributor, TableManager, DataPurger, StoreGateway},
 	},
 }
+
+// middleware for setting cache gen header to let consumer of response know all previous responses could be invalid due to delete operation
+func getHTTPCacheGenNumberHeaderSetterMiddleware(cacheGenNumbersLoader *purger.TombstonesLoader) middleware.Interface {
```

Comment:
I don't think the best way to do this is in an HTTP middleware. The routing for the queriers and query-frontend is already relatively complex. I think it would make more sense to create a new `queryable` that wraps the sharded querier.

Reply:
This is fine as a middleware and probably the only way to accomplish this right now.